README.md
# Octane Skype

Are you sick of using a subpar skype for linux that lacks even some of the basic features, such
as text search and code highlight? 

Then, it's Octane Skype to the rescue! Octane comes to leverage your skype experience with lots
of nice features!

Features:

- Unread thread count in the tray icon
- Several keyboard shortcuts
- Clicking the notifications brings the client to front
- When you focus the client, the keyboard focus is where you expect, in the chat box!
- The tray icon color reflects the user state (such as idle, busy and so on)
- Text search
- Introducing new themes
    - Compact  : Web Skype, but contacts user less space
    - Dark : Ghetto Skype's Dark Theme
    - Dark Compact: Ghetto Skype's Dark Theme, but with compact contact list.    
- Code highlight for more than 166 programming languages (using [highlight.js](https://highlightjs.org/)) with 66+ styles

Credits to [Ghetto Skype](https://github.com/stanfieldr/ghetto-skype/) for the idea of using electron and
themes. In fact, the Dark Theme is basically a copy of Ghetto's dark theme.

# Installation instructions

Ubuntu users:

```
sudo add-apt-repository ppa:g-takeuchi/octane-skype
sudo apt-get update
sudo apt-get install octane-skype
```

Other debian users, configure the following apt-get source:

```
deb http://ppa.launchpad.net/g-takeuchi/octane-skype/ubuntu wily main 
deb-src http://ppa.launchpad.net/g-takeuchi/octane-skype/ubuntu wily main 
```

And before doing apt-get, you need to add the key:

```
gpg --keyserver keyserver.ubuntu.com --recv-keys FFE10208
```

# User Guide

## Keyboard shortcuts

| Key binding     |                                     |
|-----------------|-------------------------------------|
| CTRL+G          | opens the latest unread message     |
| CTRL+F          | opens the search box                |
| CTRL+K          | focus the contact/user search field |
| ALT+Left Arrow  | Focus previous contact              |
| ALT+Right Arrow | Focus next contact                  |

## Code highlight

Code highlight only works among other Octane Users. To highlight code, start the message with three backticks ```

Regular skype users will see the backticks and the text without any format.

## Search

The search box can be used to search text that is already present in the history. So, if the history wasn't 
loaded yet, it won't be able to find it. You have to scroll up until you load more history to improve the 
search results.

# Developer Guide

## Running from the source code

Make sure you have the latest nodejs version:

```
curl -sL https://deb.nodesource.com/setup_6.x | sudo -E bash -
sudo apt-get install -y nodejs
```

Run npm to install application requirements:
```
npm install
```

On Windows, before runnning npm install, you must first install windows-build-tools:

npm install --global --production windows-build-tools

You'll also need to download and extract GTK: 

    * (32-bit) - http://ftp.gnome.org/pub/gnome/binaries/win32/gtk+/2.24/gtk+-bundle_2.24.10-20120208_win32.zip
    * (64-bit) - http://ftp.gnome.org/pub/gnome/binaries/win64/gtk+/2.22/gtk+-bundle_2.22.1-20101229_win64.zip

And extract to C:\GTK

If you get the following error during npm install:

```
error C2373: '__pfnDliNotifyHook2': redefinition; different type modifiers [...\canvas.vcxproj]
```

Run the following command and try npm install again:

```
npm -g install npm@next 
```


And to execute the application:

```
npm start
```

If you get the following error:

```
Error: ENOENT: no such file or directory, open '...\electron\path.txt'
```

Remove node_modules_ and do a npm install again.



debian/changelog
octane-skype (1.0.30-1ubuntu1~xenial) xenial; urgency=low
  
  * Added monitor to watch for communication failures, so we can at least warn the user

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Mon, 07 Nov 2016 10:35:47 -0200

octane-skype (1.0.29-1ubuntu1~wily) wily; urgency=low
  
  * Added monitor to watch for communication failures, so we can at least warn the user

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Mon, 07 Nov 2016 10:35:45 -0200

octane-skype (1.0.28-1ubuntu1~xenial) xenial; urgency=low
  
  * Fixing bug where the group pane reopens automatically

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Tue, 18 Oct 2016 07:31:07 -0200

octane-skype (1.0.27-1ubuntu1~wily) wily; urgency=low
  
  * Fixing bug where the group pane reopens automatically

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Tue, 18 Oct 2016 07:31:06 -0200

octane-skype (1.0.26-1ubuntu1~xenial) xenial; urgency=low

  * Workaround for communication issues, where skype fails to communicate with cloud.
  * Logging in the $home/.octane-skype
  * Introducing Skype Directory

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Mon, 17 Oct 2016 23:01:23 -0200

octane-skype (1.0.25-1ubuntu1~wily) wily; urgency=low

  * Workaround for communication issues, where skype fails to communicate with cloud.
  * Logging in the $home/.octane-skype
  * Introducing Skype Directory

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Mon, 17 Oct 2016 23:01:22 -0200

octane-skype (1.0.24-1ubuntu1) xenial; urgency=high

  * knockout was shipped by mistake (xenial version)

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Sat, 01 Oct 2016 15:31:41 -0300

octane-skype (1.0.23-1ubuntu1) wily; urgency=high

  * knockout was shipped by mistake

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Sat, 01 Oct 2016 15:31:41 -0300

octane-skype (1.0.22-1ubuntu1) xenial; urgency=high

  * Fixing package removing hack for webstorm (xenial version)

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Thu, 29 Sep 2016 21:44:45 -0300

octane-skype (1.0.21-1ubuntu1) wily; urgency=high

  * Fixing package removing hack for webstorm

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Thu, 29 Sep 2016 21:44:45 -0300

octane-skype (1.0.20-1ubuntu1) xenial; urgency=high

  * 1.0.19 for xenial

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Thu, 29 Sep 2016 21:44:45 -0300

octane-skype (1.0.19-1ubuntu1) wily; urgency=high

  * Join links optimization, avoiding 2 clicks to get to the group

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Thu, 29 Sep 2016 21:44:45 -0300

octane-skype (1.0.18-1ubuntu1) xenial; urgency=high

  * Xenial version for 1.0.17

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Wed, 28 Sep 2016 21:34:03 -0300


octane-skype (1.0.17-1ubuntu1) wily; urgency=high

  * Improved highlight features
  * Join links open inside electron instead of external browser

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Wed, 28 Sep 2016 21:34:03 -0300

octane-skype (1.0.16-1ubuntu1) wily; urgency=high

  * Introducing Zoom In and Out

 -- Gabriel Takeuchi <g.takeuchi@gmail.com> Sun, 25 Sep 2016 23:48:51 -0300

octane-skype (1.0.15-1ubuntu1~xenial) xenial; urgency=high

  * Finally fixing libgif issue.

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Tue, 20 Sep 2016 22:20:00 -0300


octane-skype (1.0.14-1ubuntu1) xenial; urgency=medium

  * Fixing issues with build

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sat, 17 Sep 2016 16:12:47 -0300


octane-skype (1.0.12-1ubuntu1~xenial) xenial; urgency=medium

  * Fixed libgif version for xenial

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sat, 17 Sep 2016 16:12:47 -0300

octane-skype (1.0.11-1ubuntu1) wily; urgency=medium

  * Adding explicit dependencies

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sat, 17 Sep 2016 16:12:47 -0300

octane-skype (1.0.10-1ubuntu1) wily; urgency=medium

  * Added an internal image viewer

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sat, 17 Sep 2016 16:12:47 -0300


octane-skype (1.0.9-1ubuntu1) wily; urgency=high

  * Fixed autostart
  * Fixed links without thumbnails

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sun, 11 Sep 2016 11:34:12 -0300


octane-skype (1.0.8-1ubuntu1) wily; urgency=high

  * Needs rebuild with correct node_modules

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sun, 11 Sep 2016 11:34:12 -0300

octane-skype (1.0.7-1ubuntu1) wily; urgency=high

  * Fixing images and urls open

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sun, 11 Sep 2016 11:34:12 -0300

octane-skype (1.0.6-1ubuntu1) wily; urgency=high

  * Web calls broken, fixed. Reverting arch change

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sun, 11 Sep 2016 11:34:12 -0300

octane-skype (1.0.5-1ubuntu1) wily; urgency=medium

  * Specifying architecture to prevent i386 builds (so far, unsupported)

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sun, 11 Sep 2016 11:34:12 -0300

octane-skype (1.0.4-1) wily; urgency=low

  * Testing with separate node modules

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sat, 10 Sep 2016 01:28:22 -0300

octane-skype (1.0.3-1) wily; urgency=low

  * Initial release

 -- Gabriel Takeuchi <g.takeuchi@gmail.com>  Sun, 04 Sep 2016 16:39:46 -0300


octane/trayIcon.js
const electron      = require('electron');
const Canvas        = require('canvas');
const path          = require('path');
const fs            = require('fs');
const ipcMain       = electron.ipcMain;

const nativeImage   = electron.nativeImage;
let trayIcon        = null;
let octaneInstance  = null;
let basePath        = __dirname + '/../assets/tray/';
let lastCount       = 0;
let logo            = null;

exports.initialize = function(instance) {
    octaneInstance = instance;

    var logoData = fs.readFileSync(`${basePath}skype-online.png`);
    logo = new Canvas.Image;
    logo.src = logoData;

    trayIcon   = new electron.Tray(`${basePath}skype-online.png`);
    trayIcon.setToolTip("Octane Skype");

    trayIcon.on('click', octaneInstance.toggleOpen);
    trayIcon.setContextMenu(contextMenu);
    ipcMain.on("unseen-chat-changed", (e,count) => exports.setNotificationCount(count));
    ipcMain.on("state-changed", (e,state) =>exports.updateState(state));
};

exports.setNotificationCount = function(count) {
    if (count === lastCount)
        return;

    drawTrayIcon(count);
    lastCount = count;
};

exports.updateState = function (state) {
    logo = new Canvas.Image;
    logo.src = fs.readFileSync(`${basePath}skype-${state}.png`);
    drawTrayIcon(lastCount);
};

function drawTrayIcon(count)
{
    var canvas = new Canvas(24, 24);
    var ctx = canvas.getContext('2d');

    ctx.drawImage(logo, 0, 0, logo.width, logo.height);
    if (count != 0)
        drawTrayIconMessageCount(ctx, count);

    var data = canvas.toDataURL('image/png');
    trayIcon.setImage(nativeImage.createFromDataURL(data));
}

function drawTrayIconMessageCount(ctx, count)
{
    var fontSize = 8;
    var yPos = 9;
    if (count > 99) {
        count = "∞";
        fontSize = 16;
        yPos = 11;
    }

    ctx.beginPath();
    ctx.arc(18, 6, 6, 0, 2 * Math.PI, false);
    ctx.fillStyle= 'yellow';
    ctx.fill();
    ctx.strokeStyle='#ffb31a';
    ctx.stroke();
    ctx.font = fontSize+'px Colibri';
    ctx.fillStyle = 'black';
    var tsize = ctx.measureText(count).width;
    ctx.fillText(count, 18-tsize/2, yPos);
}

let contextMenu = new electron.Menu.buildFromTemplate([
    {
        label: "Activate",
        click: () => octaneInstance.show()
    },
    {
        label: "Settings",
        click: () => octaneInstance.showSettings()
    },
    {
        label: "Online",
        icon:`${basePath}skype-online.png`,
        click: () => octaneInstance.statusChange("online")
    },
    {
        label: "Away",
        icon:`${basePath}skype-idle.png`,
        click: () => octaneInstance.statusChange("idle")
    },
    {
        label: "Busy",
        icon:`${basePath}skype-dnd.png`,
        click: () => octaneInstance.statusChange("dnd")
    },
    {
        label: "Invisible",
        icon:`${basePath}skype-hidden.png`,
        click: () => octaneInstance.statusChange("hidden")
    },
    {
        role: "quit",
        label: "Exit",
        accelerator: "Ctrl+Q",
        click: () => electron.app.quit()
    }
]);
